### ISSUE TITLE
"checkMonitoringStackHealth" timed out - disk related

### ISSUE THREAD

**Issue by Makdaam** [openshift/openshift-azure#1362]
I can't reliably reproduce this flake. Seems to be a timeout after running a long list of e2e tests (sometimes the monitoring stack doesn't come back in time to pass the test).

```
Scale Up/Down E2E tests [ScaleUpDown][Fake][LongRunning]
  Expected
      <[]*standard.TestError | len:1, cap:1>: [
          {
              Bucket: "checkMonitoringStackHealth",
              Err: {
                  s: "timed out waiting for the condition",
              },
          },
      ]
  to be empty
```
/kind test-flake

**Comment by mjudeikis** [comment]
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1361/pull-ci-openshift-openshift-azure-master-e2e-azure/3077

One more. I suspect it will underlaying disk issue.

**Comment by mjudeikis** [comment]
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1346/pull-ci-openshift-openshift-azure-master-e2e-azure/3085 
one more

**Comment by mjudeikis** [comment]
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1346/pull-ci-openshift-openshift-azure-master-upgrade-v3.2/52
Starting to think this is not a flake...

**Comment by jim-minter** [comment]
https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1361/pull-ci-openshift-openshift-azure-master-e2e-azure/3077 is not an instance of this problem!

In https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1346/pull-ci-openshift-openshift-azure-master-e2e-azure/3085, I think the following is going on:

resource group e2e-mjudeikis-1346-umcmvo, 27/3/2019 from around 13:27:31 UTC

* alertmanager-main-0 wants to start with persistent storage pvc-0fc52771-5094-11e9-ac50-000d3a173d35
* at 13:27:31 controller decides to attach disk to node; because of asynchronous API issue it believes the disk is attached when it isn't yet
* in parallel:
  1) azure disk attach takes place slowly (maybe as long as until 13:37:42?)
  2) node decides to run pod
   * node tries MountVolume.WaitForAttach 6 times between 13:27:39 and 13:28:43, but this never suceeds
   * node doesn't retry any more
   * even when disk attach finally succeeds(?), pod can't start

I think this is basically a duplicate of #1278

**Comment by jim-minter** [comment]
I think https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1346/pull-ci-openshift-openshift-azure-master-upgrade-v3.2/52 is more likely a duplicate of #1363.

**Comment by jim-minter** [comment]
[azure-activity.log](https://github.com/openshift/openshift-azure/files/3015834/azure-activity.log)
[controllers.log](https://github.com/openshift/openshift-azure/files/3015835/controllers.log)
[kubelet.log](https://github.com/openshift/openshift-azure/files/3015836/kubelet.log)

**Comment by andyzhangx** [comment]
the key error message is `Cannot find Lun for disk kubernetes-dynamic-pvc-0fc52771-5094-11e9-ac50-000d3a173d35"`, I think at that time period(13:27:39 ~13:28:43), that disk is actually not attached to the node

**Comment by jim-minter** [comment]
The kernel logs show a disk being attached at 13:27:51.  `Cannot find Lun` is still reported at 13:28:11.

Each `MountVolume.WaitForAttach entering` message is followed by a `Cannot find Lun` message, except for the `MountVolume.WaitForAttach entering` message at 13:28:43.041687.  What happens after that call?

**Comment by jim-minter** [comment]
all kernel logs for that time period:
```
2019-03-27T13:27:51 scsi 5:0:0:1: Direct-Access     Msft     Virtual Disk     1.0  PQ: 0 ANSI: 5
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] 67108864 512-byte logical blocks: (34.3 GB/32.0 GiB)
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] 4096-byte physical blocks
2019-03-27T13:27:51 sd 5:0:0:1: Attached scsi generic sg4 type 0
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] Write Protect is off
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] Mode Sense: 0f 00 10 00 
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] Write cache: disabled, read cache: enabled, supports DPO and FUA
2019-03-27T13:27:51 sd 5:0:0:1: [sdd] Attached SCSI disk
2019-03-27T13:48:31 hrtimer: interrupt took 2168766 ns
```
Looks like no attempt to mount a filesystem on that volume.

**Comment by andyzhangx** [comment]
@jim-minter  
this PR may fix your issue: https://github.com/kubernetes/kubernetes/pull/70002, detailed of this issue: https://github.com/andyzhangx/demo/blob/master/issues/azuredisk-issues.md#13-cannot-find-lun-for-disk

**Comment by mjudeikis** [comment]
We are running 1.11.1 [1]. Looks like your change is in 1.11.5+ [2] 
@jim-minter how do we address rebase in case of this scenario? 

[1] https://github.com/openshift/origin/blob/release-3.11/glide.yaml#L37
[2] https://github.com/andyzhangx/demo/blob/master/issues/azuredisk-issues.md#13-cannot-find-lun-for-disk

**Comment by asalkeld** [comment]
another one:

https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1370/pull-ci-openshift-openshift-azure-master-e2e-azure-scaleupdown/222

**Comment by jim-minter** [comment]
Backporting the kubernetes PR seems to help, going to capture kubelet goroutine dumps before applying the PR just to understand what's going on better.

`curl --cert ./cert --key key -k https://localhost:10250/debug/pprof/goroutine?debug=2`

**Comment by jim-minter** [comment]
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1397/pull-ci-openshift-openshift-azure-master-upgrade-v3.2/229/build-log.txt

is another example of this

**Comment by jim-minter** [comment]
[kubelet.log](https://github.com/openshift/openshift-azure/files/3037041/kubelet.log)

@andyzhangx in the log attached here, the `Cannot find Lun` message happens exactly 15 minutes (02:12:30) after the last `MountVolume.WaitForAttach` message.  What can cause a timeout of exactly 15 minutes in this code path?

I am curious about:
```golang
	availabilitySetNodesCacheTTL      = 15 * time.Minute
	nodeNameToScaleSetMappingCacheTTL = 15 * time.Minute
```

but I can't work out how that would cause this code path to be delayed?

**Comment by andyzhangx** [comment]
@jim-minter that's correct, for VMSS, at that version, it's 15 minutes, now it's changed to 5 minutes, and this PR: https://github.com/kubernetes/kubernetes/pull/70002 could fix the issue by finding the lun directly from path like `/dev/disk/azure/lunX`

**Comment by jim-minter** [comment]
@andyzhangx I want to understand what the underlying issue is - what is the actual code path that is hanging for 15 minutes?

@feiskyer do you know what the path could be?

**Comment by andyzhangx** [comment]
@jim-minter I was wrong, the `waitForAttachTimeout` value is defined here, it's 10 min:
https://github.com/kubernetes/kubernetes/blob/51db0fe21a816b0818df571ad70386864d83c722/pkg/kubelet/volumemanager/volume_manager.go#L80-L86

And  Azure disk `WaitForAttach` code is here:
https://github.com/kubernetes/kubernetes/blob/51db0fe21a816b0818df571ad70386864d83c722/pkg/volume/azure_dd/attacher.go#L194

And PR(https://github.com/kubernetes/kubernetes/pull/70002) could fix that issue.

I am not sure why that disk is not displayed under `/dev/disk/azure/lun0` after 15 min, is it still on VMSS? and what's your k8s version?

**Comment by andyzhangx** [comment]
actually all the VM cache(VMSS and VMAS) holds only 1 min, I don't know even after 15 min, why `/dev/disk/azure/lun0` is not created in the VM although disk attachment succeeded.

```
I0403 01:57:30.643787    6096 operation_generator.go:489] MountVolume.WaitForAttach entering for volume "pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89" (UniqueName: "kubernetes.io/azure-disk//subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/e2e-upgrade-jim-minter-1397-bpxojd/providers/Microsoft.Compute/disks/kubernetes-dynamic-pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89") pod "postgresql-1-6w5g2" (UID: "d4c5a3f8-55b3-11e9-b9ab-000d3a52dd89") DevicePath "0"
...
E0403 02:12:30.710683    6096 nestedpendingoperations.go:267] Operation for "\"kubernetes.io/azure-disk//subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/e2e-upgrade-jim-minter-1397-bpxojd/providers/Microsoft.Compute/disks/kubernetes-dynamic-pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89\"" failed. No retries permitted until 2019-04-03 02:12:31.210619929 +0000 UTC m=+1327.831889311 (durationBeforeRetry 500ms). Error: "MountVolume.WaitForAttach failed for volume \"pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89\" (UniqueName: \"kubernetes.io/azure-disk//subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/resourceGroups/e2e-upgrade-jim-minter-1397-bpxojd/providers/Microsoft.Compute/disks/kubernetes-dynamic-pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89\") pod \"postgresql-1-6w5g2\" (UID: \"d4c5a3f8-55b3-11e9-b9ab-000d3a52dd89\") : Cannot find Lun for disk kubernetes-dynamic-pvc-95fdb811-55b3-11e9-b9ab-000d3a52dd89"
```

**Comment by andyzhangx** [comment]
@jim-minter could you paste the  udev rule `/etc/udev/rules.d/66-azure-storage.rules` on that VM:
it could be related to this issue whch is also related to udev rule missing:
https://github.com/kubernetes/kubernetes/issues/66443#issuecomment-406765240

**Comment by feiskyer** [comment]
> @feiskyer do you know what the path could be?

Not sure from current findings. Is it possible to increase the log level of kubelet, so that more details could be found?

**Comment by jim-minter** [comment]
@feiskyer @andyzhangx  I've now got a kubelet goroutine dump at the point where this is happening.

https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1430/pull-ci-openshift-openshift-azure-master-e2e-azure/3415/artifacts/internal-e2e/compute-testapp-goroutine-dump

Goroutine 219 is cloudResourceSyncManager -> NodeAddresses -> getVmssVM.  It has been stuck for 10 minutes because of 429(s?) returned by Azure.

Goroutine 3792 is GetDiskLun.  It is stuck on Get because 219 holds the mutex.

This explains why the patch helps - by removing GetDiskLun->getVmssVM we can't get stuck in this path.

It is problematic that getVmssVM could get stuck for 10 minutes due to 429s without logging anything.

Maybe we are not so likely to hit this in production: in prod, each cluster has its own service principals.  In test, multiple clusters share the same service principals so rate limiting is more problematic.

I'm inclined to proceed by:
1) lower priority backport of https://github.com/kubernetes/kubernetes/pull/70002 to improve robustness
2) further refactor our use of service principals in test

**Comment by jim-minter** [comment]
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_openshift-azure/1437/pull-ci-openshift-openshift-azure-master-upgrade-v3.2/344/build-log.txt

The above is a log where the *controller* gets stuck in AttachVolume.Attach for 15 minutes.  I bet it's a similar issue, but this one would not be fixed by kubernetes/kubernetes#70002.

Enabling UseInstanceMetadata is another key action here, that will radically reduce our Azure API call rate.

**Comment by jim-minter** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1696901

**Comment by andyzhangx** [comment]
@jim-minter thanks for the detailed analysis, 429 is `StatusTooManyRequests`, and PR(https://github.com/kubernetes/kubernetes/pull/70002) fixed this issue on following k8s versions:

| k8s version | fixed version |
| ---- | ---- |
| v1.9 | no such issue |
| v1.10 | 1.10.10 |
| v1.11 | 1.11.5 |
| v1.12 | 1.12.3 |
| v1.13 | no such issue |

Which k8s version are you testing on OpenShift then?

**Comment by jim-minter** [comment]
@andyzhangx this is OpenShift 3.11 which is k8s 1.11.0 + selected backports.  It does not include kubernetes/kubernetes#70002 but we will add it.

**Comment by feiskyer** [comment]
@jim-minter let's know if there're still issues.

**Comment by jim-minter** [comment]
error not seen since 08/04/19 (mainly due to separating out service principals), closing
backport taking place at https://bugzilla.redhat.com/show_bug.cgi?id=1696901
@feiskyer @andyzhangx many thanks for all your help with this.
